inc/api/namespace.php

```diff
@@ -223,285 +205,146 @@ function get_query( array $events, int $start, int $end, array $aggs, ?Filter $f
-				'must_not' => [],
-				'should' => [],
-			],
-		],
-		'size' => 0,
-		'aggs' => $aggs,
-	];
-
-	return $query;
-}
-
-/**
- * Get global page view stats.
- *
- * @param int $start The start timestamp.
- * @param int $end The end timestamp.
- * @param string|int $resolution Resolution for histogram data.
- * @param Filter|null $filter Query filter object.
- * @return array|WP_error
- */
-function get_graph_data( $start, $end, $resolution = 'day', ?Filter $filter = null ) {
-	$returning = [
-		'filter' => [
-			'range' => [
-				'endpoint.Metrics.sessions' => [
-					'gt' => 1,
-				],
-			],
-		],
-		'aggregations' => [
-			'user' => [
-				'cardinality' => [
-					'field' => 'endpoint.Id.keyword',
-				],
-			],
-		],
-	];
-
-	$bounce = [
-		'range' => [
-			'field' => 'endpoint.Metrics.pageViews',
-			'keyed' => true,
-			'ranges' => [
-				[
-					'key' => 'single',
-					'to' => 2,
-				],
-				[
-					'key' => 'multiple',
-					'from' => 2,
-				],
-			],
-		],
-		'aggregations' => [
-			'user' => [
-				'cardinality' => [
-					'field' => 'endpoint.Id.keyword',
-				],
-			],
-		],
-	];
-
-	$aggs = [
-		'views' => [
-			'value_count' => [
-				'field' => 'event_timestamp',
-			],
-		],
-		'visitors' => [
-			'cardinality' => [
-				'field' => 'endpoint.Id.keyword',
-			],
-		],
-		'returning' => $returning,
-		'bounce' => $bounce,
-		'by_date_and_user' => [
-			'date_histogram' => [
-				'field' => 'event_timestamp',
-				'interval' => $resolution,
-				'extended_bounds' => [
-					'min' => (int) sprintf( '%d000', $start ),
-					'max' => (int) sprintf( '%d999', $end ),
-				],
-			],
-			'aggregations' => [
-				'user' => [
-					'cardinality' => [
-						'field' => 'endpoint.Id.keyword',
-					],
-				],
-				'returning' => $returning,
-				'bounce' => $bounce,
-			],
-		],
-		'by_date_bucket' => [
-			'date_histogram' => [
-				'field' => 'event_timestamp',
-				'interval' => $resolution,
-				'extended_bounds' => [
-					'min' => (int) sprintf( '%d000', $start ),
-					'max' => (int) sprintf( '%d999', $end ),
-				],
-			],
-		],
-	];
-
-	$query = get_query( [ 'pageView' ], $start, $end, $aggs, $filter );
-
-	// Add registered term aggregations.
-	$registered_aggs = apply_filters( 'altis.analytics.ui.graph_aggregations', [] );
-	foreach ( $registered_aggs as $name => $agg_options ) {
-		$query['aggs'][ $name ] = $agg_options['aggregation'];
-	}
+	$query_where = sprintf( 'AND (%s)', implode( ') AND (', $query_where ) );
+
+	// Build SQL query.
+	$query = $wpdb->prepare(
+		"SELECT
+			toUnixTimestamp(toStartOfInterval(event_timestamp, INTERVAL {$resolution})) as `date`,
+			-- views
+			count() as views,
+			-- uniques
+			uniqCombined64(endpoint_id) as uniques,
+			-- returning
+			uniqCombined64If(endpoint_id, endpoint_metrics['sessions'] > 1) as returning,
+			-- bounce
+			uniqCombined64If(endpoint_id, endpoint_metrics['pageViews'] < 2) as bounce
+		FROM analytics
+		WHERE
+			attributes['blogId'] = %s
```

Comment:
Does Clickhouse support parameterised queries? That'd be better than building a raw query

Reply:
Appears so, wasn't aware of that feature https://clickhouse.com/docs/en/interfaces/http/#cli-queries-with-parameters

Reply:
@rmccue I'll do a follow up to parameterise the queries.